src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+                {
+                    buf.Append(-remaining)
+                        .Append(" ago");
+                }
+                else
+                {
+                    buf.Append("now");
+                }
+
+                if (this.Cancelled)
+                {
+                    buf.Append(", cancelled");
+                }
+
+                return buf.Append(", task: ")
+                    .Append(this.Task)
+                    .Append(')')
+                    .ToString();
+            }
+        }
+
+        /// <summary>
+        /// Bucket that stores HashedWheelTimeouts. These are stored in a linked-list like datastructure to allow easy
+        /// removal of HashedWheelTimeouts in the middle. Also the HashedWheelTimeout act as nodes themself and so no
+        /// extra object creation is needed.
+        /// </summary>
+        sealed class HashedWheelBucket
+        {
+            // Used for the linked-list datastructure
+            HashedWheelTimeout head;
+            HashedWheelTimeout tail;
+
+            /// <summary>
+            /// Add {@link HashedWheelTimeout} to this bucket.
+            /// </summary>
+            public void AddTimeout(HashedWheelTimeout timeout)
+            {
+                Contract.Assert(timeout.Bucket == null);
+                timeout.Bucket = this;
+                if (this.head == null)
+                {
+                    this.head = this.tail = timeout;
+                }
+                else
+                {
+                    this.tail.Next = timeout;
+                    timeout.Prev = this.tail;
+                    this.tail = timeout;
+                }
+            }
+
+            /// <summary>
+            /// Expire all <see cref="HashedWheelTimeout"/>s for the given <c>deadline</c>.
+            /// </summary>
+            public void ExpireTimeouts(TimeSpan deadline)
+            {
+                HashedWheelTimeout timeout = this.head;
+
+                // process all timeouts
+                while (timeout != null)
+                {
+                    HashedWheelTimeout next = timeout.Next;
+                    if (timeout.RemainingRounds <= 0)
+                    {
+                        next = this.Remove(timeout);
+                        if (timeout.Deadline <= deadline)
+                        {
+                            timeout.Expire();
+                        }
+                        else
+                        {
+                            // The timeout was placed into a wrong slot. This should never happen.
+                            throw new InvalidOperationException(
+                                string.Format(
+                                    "timeout.deadline (%d) > deadline (%d)",
+                                    timeout.Deadline,
+                                    deadline));
+                        }
+                    }
+                    else if (timeout.Cancelled)
```

Comment:
don't do this from two places. you already have a place where you are reading off of the canceled queue and removing it at that point. otherwise Remove() is called twice for the same timeout object potentially and from what I can tell it can lead to weird behavior (such as double decrementing pending timer count)